File: async.lua
local ffi = require('ffi')
local util = require('util')
local sched = require('sched')
local pthread = require('pthread')
local trigger = require('trigger')
local inspect = require('inspect')
local adt = require('adt')

ffi.cdef [[

typedef void (*zz_async_handler)(void *request_data);

int zz_async_register_worker(void *handlers[]);

struct zz_async_worker_info {
  zz_trigger request_trigger;
  int worker_id;
  int handler_id;
  void *request_data;
  zz_trigger response_trigger;
};

void *zz_async_worker_thread(void *arg);

enum {
  ZZ_ASYNC_ECHO
};

struct zz_async_echo {
  double delay;
  double payload;
  double response;
};

void *zz_async_handlers[];

]]

local M = {}

local MAX_ACTIVE_THREADS = 16

-- pool of free (reservable) worker threads
local thread_pool = {}

-- reservation ids of coroutines waiting for a worker thread
local reserve_queue = adt.List()

-- number of worker threads (free + active)
local n_worker_threads   = 0

-- number of worker threads currently servicing a request
local n_active_threads   = 0

local function create_worker_thread()
   n_worker_threads = n_worker_threads + 1
   local worker_info = ffi.new("struct zz_async_worker_info")
   local request_trigger = trigger()
   worker_info.request_trigger = request_trigger
   local response_trigger = trigger()
   worker_info.response_trigger = response_trigger
   sched.poller_add(response_trigger.fd, "r")
   local thread_id = ffi.new("pthread_t[1]")
   local rv = ffi.C.pthread_create(thread_id,
                                   nil,
                                   ffi.C.zz_async_worker_thread,
                                   ffi.cast("void*", worker_info))
   if rv ~= 0 then
      error("cannot create async worker thread: pthread_create() failed")
   end
   local self = {}
   function self:send_request(worker_id, handler_id, request_data)
      worker_info.worker_id = worker_id
      worker_info.handler_id = handler_id
      worker_info.request_data = request_data
      request_trigger:fire()
      response_trigger:poll()
   end
   function self:stop()
      self:send_request(-1, 0, nil)
      local retval = ffi.new("void*[1]")
      local rv = ffi.C.pthread_join(thread_id[0], retval)
      if rv ~=0 then
         error("cannot join async worker thread: pthread_join() failed")
      end
      sched.poller_del(response_trigger.fd)
      response_trigger:delete()
      request_trigger:delete()
      n_worker_threads = n_worker_threads - 1
   end
   return self
end

local function reserve_thread()
   local t
   if #thread_pool == 0 then
      if n_active_threads == MAX_ACTIVE_THREADS then
         local reservation_id = sched.make_event_id()
         reserve_queue:push(reservation_id)
         -- block until we get a free thread
         t = sched.wait(reservation_id)
      else
         t = create_worker_thread()
      end
   else
      t = table.remove(thread_pool)
   end
   n_active_threads = n_active_threads + 1
   return t
end

local function release_thread(t)
   n_active_threads = n_active_threads - 1
   if reserve_queue:empty() then
      -- nobody is waiting for a thread, put it back into the pool
      table.insert(thread_pool, t)
   else
      local reservation_id = reserve_queue:shift()
      sched.emit(reservation_id, t)
   end
end

function M.register_worker(handlers)
   return ffi.C.zz_async_register_worker(handlers)
end

function M.request(worker_id, handler_id, request_data)
   -- reserve_thread() blocks if needed
   -- until a thread becomes available
   local t = reserve_thread()
   t:send_request(worker_id, handler_id, request_data)
   release_thread(t)
end

local function AsyncModule(sched)
   local self = {}
   function self.init()
      thread_pool = {}
      reserve_queue = adt.List()
      n_active_threads = 0
      n_worker_threads = 0
   end
   function self.done()
      assert(n_active_threads == 0)
      assert(reserve_queue:empty())
      for _,t in ipairs(thread_pool) do
         t:stop()
      end
      assert(n_worker_threads == 0)
      thread_pool = {}
   end
   return self
end

sched.register_module(AsyncModule)

return M

File: async.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <math.h>
#include <poll.h>

#include "trigger.h"

typedef void (*zz_async_handler)(void *request_data);

#define MAX_REGISTERED_WORKERS 256

struct zz_async_worker {
  int handler_count;
  zz_async_handler *handlers;
};

static struct zz_async_worker registered_workers[MAX_REGISTERED_WORKERS];
static int registered_worker_count = 0;

int zz_async_register_worker(void *handlers[]) {
  if (registered_worker_count == MAX_REGISTERED_WORKERS) {
    fprintf(stderr, "async: cannot register more workers, %d limit exceeded\n",
            MAX_REGISTERED_WORKERS);
    exit(1);
  }
  struct zz_async_worker *z = &registered_workers[registered_worker_count++];
  z->handlers = (zz_async_handler *) handlers;
  z->handler_count = 0;
  // list of handlers must be NULL-terminated
  while (*handlers != NULL) {
    z->handler_count++;
    handlers++;
  }
  // worker id is 1-based
  return registered_worker_count;
}

struct zz_async_worker_info {
  zz_trigger request_trigger;
  int worker_id;
  int handler_id;
  void *request_data;
  zz_trigger response_trigger;
};

void *zz_async_worker_thread(void *arg) {
  struct zz_async_worker_info *info = (struct zz_async_worker_info*) arg;

  /* When the Lua side wants to execute something which would block
   * the event loop, it posts it as a request to one of the async
   * worker threads.
   *
   * Every module can define its request types in C code and register
   * the corresponding handlers with the async module by calling
   * zz_async_register_worker(). A worker is a group of handlers
   * registered in this way. Worker threads can be asked to execute
   * any handler provided by a registered worker.
   *
   * To make a request, the Lua side fills out the selected worker
   * thread's zz_async_worker_info structure with the worker_id,
   * handler_id and a pointer to a structure describing the request
   * (the layout varies by request type). The Lua side then fires the
   * worker thread's request_trigger - this wakes up the thread which
   * then looks up the desired handler and passes request_data to it.
   *
   * Before the request handler completes, it should store any return
   * values in the request_data structure. The worker thread then
   * fires response_trigger which is being polled on the Lua side in
   * the coroutine which is waiting for the completion of the async
   * request.
   */

  while (1) {
    zz_trigger_wait(&info->request_trigger);
    /* worker_id is a 1-based index to the registered_workers array */
    int worker_id = info->worker_id;
    /* worker_id == -1 is the exit signal */
    if (worker_id == -1) {
      zz_trigger_fire(&info->response_trigger);
      break;
    }
    if (worker_id < 1 || worker_id > registered_worker_count) {
      fprintf(stderr, "invalid async request: worker_id is out of range (registered_worker_count=%d, worker_id=%d)\n", registered_worker_count, worker_id);
      exit(1);
    }
    struct zz_async_worker *worker = &registered_workers[worker_id-1];
    /* handler id identifies the handler within the selected worker */
    int handler_id = info->handler_id;
    /* handler_id is 0-based */
    if (handler_id < 0 || handler_id >= worker->handler_count) {
      fprintf(stderr, "invalid async request: handler_id is out of range (worker_id=%d, handler_id=%u, handler_count=%d)\n", worker_id, handler_id, worker->handler_count);
      exit(1);
    }
    zz_async_handler handler = worker->handlers[handler_id];
    handler(info->request_data); /* process request */
    zz_trigger_fire(&info->response_trigger);
  }
  return NULL;
}

/* a pre-defined handler for testing purposes */

enum {
  ZZ_ASYNC_ECHO
};

struct zz_async_echo {
  double delay;
  double payload;
  double response;
};

void zz_async_echo(struct zz_async_echo *r) {
  double fractional_part, integer_part;
  fractional_part = modf(r->delay, &integer_part);
  struct timespec rqtp;
  rqtp.tv_sec = integer_part;
  rqtp.tv_nsec = fractional_part * 1e9;
  if (nanosleep(&rqtp, NULL) != 0) {
    fprintf(stderr, "zz_async_echo: nanosleep() failed\n");
  }
  r->response = r->payload;
}

void *zz_async_handlers[] = {
  zz_async_echo,
  0
};
